Index: vehicle.c
===================================================================
--- vehicle.c	(revision 2840)
+++ vehicle.c	(working copy)
@@ -58,12 +58,15 @@
 
 bool VehicleNeedsService(const Vehicle *v)
 {
+	if (v->vehstatus & VS_CRASHED)
+		return false; /* Crashed vehicles don't need service anymore */
+
+	if (GetPlayer(v->owner)->engine_replacement[v->engine_type] != INVALID_ENGINE)
+		return true; /* Vehicle is due to be replaced */
+
 	if (_patches.no_servicing_if_no_breakdowns && _opt.diff.vehicle_breakdowns == 0)
 		return false;
 
-	if (v->vehstatus & VS_CRASHED)
-		return false; /* Crashed vehicles don't need service anymore */
-
 	return _patches.servint_ispercent ?
 		(v->reliability < GetEngine(v->engine_type)->reliability * (100 - v->service_interval) / 100) :
 		(v->date_of_last_service + v->service_interval < _date);
