Index: ttd.c
===================================================================
--- ttd.c	(revision 2261)
+++ ttd.c	(working copy)
@@ -62,6 +62,8 @@
 bool disable_computer; // We should get ride of this thing.. is only used for a debug-cheat
 static byte _os_version = 0;
 
+static uint16 _num_sign_sort; // Sign sort thing
+
 /* TODO: usrerror() for errors which are not of an internal nature but
  * caused by the user, i.e. missing files or fatal configuration errors.
  * Post-0.4.0 since Celestar doesn't want this in SVN before. --pasky */
@@ -461,6 +463,7 @@
 static void LoadIntroGame(void)
 {
 	char filename[256];
+	const SignStruct *ss;
 
 	_game_mode = GM_MENU;
 	CLRBITS(_display_opt, DO_TRANS_BUILDINGS); // don't make buildings transparent in intro
@@ -483,6 +486,21 @@
 			GenerateWorld(1, 6, 6); // if failed loading, make empty world.
 	}
 
+	// Sort signs for movey thingy
+	_num_sign_sort = 0;
+
+	_sign_sort = realloc(_sign_sort, GetSignPoolSize() * sizeof(_sign_sort[0]));
+	if (_sign_sort == NULL)
+		error("Could not allocate memory for the sign-sorting-list");
+
+	FOR_ALL_SIGNS(ss) {
+		if(ss->str != STR_NULL) {
+			_sign_sort[_num_sign_sort++] = ss->index;
+		}
+	}
+
+	_sign_last_update = 0;
+
 	_pause = 0;
 	_local_player = 0;
 	MarkWholeScreenDirty();
@@ -1025,6 +1043,33 @@
 		_current_player = p;
 	}
 
+	if (_game_mode == GM_MENU && _tick_counter - _sign_last_update > 370) {
+		// 370 ticks ~ 10 secs
+		SignStruct *ss = NULL;
+		Window *w = FindWindowById(WC_MAIN_WINDOW, 0);
+		char buf1[64];
+		int zoom_level;
+		assert(w);
+
+		if (_num_sign_sort > 0) {
+			_view_sign_number++;
+			if (_view_sign_number >= _num_sign_sort)
+				_view_sign_number = 0;
+
+			ss = GetSign(_sign_sort[_view_sign_number]);
+			GetString(buf1, ss->str);
+			zoom_level = strtol(buf1, NULL, 10);
+			if (zoom_level < 0 || zoom_level > 2)
+				zoom_level = 0;
+		
+			DEBUG(misc, 2) ("Sign number: %d Zoom: %d", ss->index, zoom_level);
+			w->viewport->zoom = zoom_level;
+	         	ScrollMainWindowToTile(TILE_FROM_XY(ss->x, ss->y));
+		}
+
+		_sign_last_update = _tick_counter;
+	}
+
 	_in_state_game_loop = false;
 }
 
Index: ttd.h
===================================================================
--- ttd.h	(revision 2261)
+++ ttd.h	(working copy)
@@ -536,6 +536,9 @@
 };
 VARDEF byte _no_scroll;
 
+VARDEF uint16 _sign_last_update;
+VARDEF uint32 _view_sign_number;
+
 #include "functions.h"
 #include "variables.h"
 
Index: viewport.c
===================================================================
--- viewport.c	(revision 2261)
+++ viewport.c	(working copy)
@@ -916,7 +916,7 @@
 	int left, top, right, bottom;
 	StringSpriteToDraw *sstd;
 
-	if (!(_display_opt & DO_SHOW_SIGNS))
+	if (!(_display_opt & DO_SHOW_SIGNS || _game_mode == GM_MENU))
 		return;
 
 	left = dpi->left;
