Index: smallmap_gui.c
===================================================================
--- smallmap_gui.c	(revision 2361)
+++ smallmap_gui.c	(working copy)
@@ -12,6 +12,7 @@
 #include "vehicle.h"
 #include "town.h"
 #include "sound.h"
+#include "debug.h"
 
 static const Widget _smallmap_widgets[] = {
 {    WWT_TEXTBTN,   RESIZE_NONE,    13,     0,    10,     0,    13, STR_00C5,		STR_018B_CLOSE_WINDOW},
@@ -25,9 +26,11 @@
 {     WWT_IMGBTN,   RESIZE_LRTB,    13,   380,   401,   280,   301, 0x2E4,				STR_0194_SHOW_TRANSPORT_ROUTES_ON},
 {     WWT_IMGBTN,   RESIZE_LRTB,    13,   402,   423,   280,   301, 0x2E6,				STR_0195_SHOW_VEGETATION_ON_MAP},
 {     WWT_IMGBTN,   RESIZE_LRTB,    13,   424,   445,   280,   301, 0x2E7,				STR_0196_SHOW_LAND_OWNERS_ON_MAP},
-{     WWT_IMGBTN,   RESIZE_LRTB,    13,   358,   379,   258,   279, 0x0,					STR_NULL},
+{     WWT_PANEL,   RESIZE_LRTB,    13,   358,   379,   258,   279, 0x0,					STR_NULL},
 {     WWT_IMGBTN,   RESIZE_LRTB,    13,   358,   379,   280,   301, 0xFED,				STR_0197_TOGGLE_TOWN_NAMES_ON_OFF},
-{     WWT_IMGBTN,    RESIZE_RTB,    13,     0,   357,   258,   301, 0x0,					STR_NULL},
+{     WWT_PANEL,   RESIZE_LRTB,    13,   336,   357,   258,   279, 0x2DF,					STR_NULL},
+{     WWT_PANEL,   RESIZE_LRTB,    13,   336,   357,   280,   301, 0x2E0,					STR_NULL},
+{     WWT_IMGBTN,    RESIZE_RTB,    13,     0,   335,   258,   301, 0x0,					STR_NULL},
 {      WWT_PANEL,    RESIZE_RTB,    13,     0,   433,   302,   313, 0x0,					STR_NULL},
 {  WWT_RESIZEBOX,   RESIZE_LRTB,    13,   434,   445,   302,   313, 0x0,					STR_RESIZE_BUTTON},
 {   WIDGETS_END},
@@ -35,6 +38,7 @@
 
 static int _smallmap_type;
 static bool _smallmap_show_towns = true;
+static int _smallmap_zoom = 1;
 
 #define MK(a,b) a,b
 #define MKEND() 0xffff
@@ -335,11 +339,12 @@
 
 static void DrawSmallMapContours(byte *dst, uint xc, uint yc, int pitch, int reps, uint32 mask)
 {
+	reps--;
 	do {
-		if (xc < MapMaxX() && yc < MapMaxY())
+		if (xc*_smallmap_zoom < MapMaxX() && yc*_smallmap_zoom < MapMaxY())
 			if (dst > _screen.dst_ptr && dst < (_screen.dst_ptr + _screen.width * _screen.height - _screen.width))
-				WRITE_PIXELS_OR(dst, GetSmallMapContoursPixels(TILE_XY(xc,yc)) & mask);
-	} while (xc++,yc++,dst+=pitch,--reps != 0);
+				WRITE_PIXELS_OR(dst, GetSmallMapContoursPixels(TILE_XY(xc*_smallmap_zoom,yc*_smallmap_zoom)) & mask);
+	} while (xc++,yc++,dst+=pitch,reps-- != 0);
 }
 
 
@@ -354,8 +359,8 @@
 static void DrawSmallMapVehicles(byte *dst, uint xc, uint yc, int pitch, int reps, uint32 mask)
 {
 	do {
-		if (xc < MapMaxX() && yc < MapMaxY())
-			WRITE_PIXELS_OR( dst, GetSmallMapVehiclesPixels(TILE_XY(xc,yc)) & mask );
+		if (xc*_smallmap_zoom < MapMaxX() && yc*_smallmap_zoom < MapMaxY())
+			WRITE_PIXELS_OR(dst, GetSmallMapVehiclesPixels(TILE_XY(xc*_smallmap_zoom,yc*_smallmap_zoom)) & mask);
 	} while (xc++,yc++,dst+=pitch,--reps != 0);
 }
 
@@ -399,8 +404,8 @@
 static void DrawSmallMapIndustries(byte *dst, uint xc, uint yc, int pitch, int reps, uint32 mask)
 {
 	do {
-		if (xc < MapMaxX() && yc < MapMaxY())
-			WRITE_PIXELS_OR(dst, GetSmallMapIndustriesPixels(TILE_XY(xc,yc)) & mask);
+		if (xc*_smallmap_zoom < MapMaxX() && yc*_smallmap_zoom < MapMaxY())
+			WRITE_PIXELS_OR(dst, GetSmallMapIndustriesPixels(TILE_XY(xc*_smallmap_zoom,yc*_smallmap_zoom)) & mask);
 	} while (xc++,yc++,dst+=pitch,--reps != 0);
 }
 
@@ -428,8 +433,8 @@
 static void DrawSmallMapRoutes(byte *dst, uint xc, uint yc, int pitch, int reps, uint32 mask)
 {
 	do {
-		if (xc < MapMaxX() && yc < MapMaxY())
-			WRITE_PIXELS_OR(dst, GetSmallMapRoutesPixels(TILE_XY(xc,yc)) & mask);
+		if (xc*_smallmap_zoom < MapMaxX() && yc*_smallmap_zoom < MapMaxY())
+			WRITE_PIXELS_OR(dst, GetSmallMapRoutesPixels(TILE_XY(xc*_smallmap_zoom,yc*_smallmap_zoom)) & mask);
 	} while (xc++,yc++,dst+=pitch,--reps != 0);
 }
 
@@ -484,8 +489,8 @@
 static void DrawSmallMapVegetation(byte *dst, uint xc, uint yc, int pitch, int reps, uint32 mask)
 {
 	do {
-		if (xc < MapMaxX() && yc < MapMaxY())
-			WRITE_PIXELS_OR(dst, GetSmallMapVegetationPixels(TILE_XY(xc,yc)) & mask);
+		if (xc*_smallmap_zoom < MapMaxX() && yc*_smallmap_zoom < MapMaxY())
+			WRITE_PIXELS_OR(dst, GetSmallMapVegetationPixels(TILE_XY(xc*_smallmap_zoom,yc*_smallmap_zoom)) & mask);
 	} while (xc++,yc++,dst+=pitch,--reps != 0);
 }
 
@@ -511,8 +516,8 @@
 static void DrawSmallMapOwners(byte *dst, uint xc, uint yc, int pitch, int reps, uint32 mask)
 {
 	do {
-		if (xc < MapMaxX() && yc < MapMaxY())
-			WRITE_PIXELS_OR(dst, GetSmallMapOwnerPixels(TILE_XY(xc,yc)) & mask);
+		if (xc*_smallmap_zoom < MapMaxX() && yc*_smallmap_zoom < MapMaxY())
+			WRITE_PIXELS_OR(dst, GetSmallMapOwnerPixels(TILE_XY(xc*_smallmap_zoom,yc*_smallmap_zoom)) & mask);
 	} while (xc++,yc++,dst+=pitch,--reps != 0);
 }
 
@@ -681,8 +686,8 @@
 					(v->vehstatus & (VS_HIDDEN | VS_UNCLICKABLE)) == 0) {
 				// Remap into flat coordinates.
 				Point pt = RemapCoords(
-					(v->x_pos - WP(w,smallmap_d).scroll_x) / 16,
-					(v->y_pos - WP(w,smallmap_d).scroll_y) / 16,
+					((v->x_pos / _smallmap_zoom) - WP(w,smallmap_d).scroll_x) / 16,
+					((v->y_pos / _smallmap_zoom)  - WP(w,smallmap_d).scroll_y) / 16,
 					0);
 				x = pt.x;
 				y = pt.y;
@@ -729,8 +734,8 @@
 			if (t->xy != 0) {
 				// Remap the town coordinate
 				Point pt = RemapCoords(
-					(int)(TileX(t->xy) * 16 - WP(w, smallmap_d).scroll_x) / 16,
-					(int)(TileY(t->xy) * 16 - WP(w, smallmap_d).scroll_y) / 16,
+					(int)((TileX(t->xy)/_smallmap_zoom) * 16 - WP(w, smallmap_d).scroll_x) / 16,
+					(int)((TileY(t->xy)/_smallmap_zoom) * 16 - WP(w, smallmap_d).scroll_y) / 16,
 					0);
 				x = pt.x - WP(w,smallmap_d).subscroll + 3 - (t->sign.width_2 >> 1);
 				y = pt.y;
@@ -757,10 +762,10 @@
 
 		pt = RemapCoords(WP(w, smallmap_d).scroll_x, WP(w, smallmap_d).scroll_y, 0);
 
-		x = vp->virtual_left - pt.x;
-		y = vp->virtual_top - pt.y;
-		x2 = (x + vp->virtual_width) / 16;
-		y2 = (y + vp->virtual_height) / 16;
+		x = (vp->virtual_left / _smallmap_zoom) - pt.x;
+		y = (vp->virtual_top / _smallmap_zoom) - pt.y;
+		x2 = (x + (vp->virtual_width / _smallmap_zoom)) / 16;
+		y2 = (y + (vp->virtual_height / _smallmap_zoom)) / 16;
 		x /= 16;
 		y /= 16;
 
@@ -850,6 +855,18 @@
 			SetWindowDirty(w);
 			SndPlayFx(SND_15_BEEP);
 			break;
+		case 13: /* zoom in */
+			if (_smallmap_zoom > 1)
+				_smallmap_zoom--;
+			SetWindowDirty(w);
+			SndPlayFx(SND_15_BEEP);
+			break;
+		case 14: /* zoom out */
+			if (_smallmap_zoom < 8)
+				_smallmap_zoom++;
+			SetWindowDirty(w);
+			SndPlayFx(SND_15_BEEP);
+			break;
 		}
 		break;
 
