Index: sound.c
===================================================================
--- sound.c	(revision 2900)
+++ sound.c	(working copy)
@@ -23,6 +23,8 @@
 static FileEntry* _files;
 
 #define SOUND_SLOT 31
+#define PANNING_LEVELS 16
+#define PANNING_LEVELS_SHIFT 3
 
 
 static void OpenBankFile(const char *filename)
@@ -132,15 +134,18 @@
 }
 
 // Low level sound player
-static void StartSound(uint sound, uint panning, uint volume)
+static void StartSound(uint sound, int panning, uint volume)
 {
 	MixerChannel* mc;
+	uint left_vol, right_vol;
 
 	if (volume == 0) return;
 	mc = MxAllocateChannel(_mixer);
 	if (mc == NULL) return;
 	if (!SetBankSource(mc, sound)) return;
-	MxSetChannelVolume(mc, volume << 8, volume << 8);
+	left_vol = (volume * PANNING_LEVELS) - (volume * panning);
+	right_vol = (volume * PANNING_LEVELS) + (volume * panning);
+	MxSetChannelVolume(mc, left_vol << PANNING_LEVELS_SHIFT, right_vol << PANNING_LEVELS_SHIFT);
 	MxActivateChannel(mc);
 }
 
@@ -185,11 +190,11 @@
 		if (vp != NULL &&
 				IS_INSIDE_1D(x, vp->virtual_left, vp->virtual_width) &&
 				IS_INSIDE_1D(y, vp->virtual_top, vp->virtual_height)) {
-			int left = ((x - vp->virtual_left) >> vp->zoom) + vp->left;
+			int left = (x - vp->virtual_left);
 
 			StartSound(
 				_sound_idx[sound],
-				clamp(left / 71, 0, 8),
+				clamp(left / (vp->virtual_width / ((PANNING_LEVELS << 1) + 1)) - PANNING_LEVELS, -PANNING_LEVELS, PANNING_LEVELS),
 				(_sound_base_vol[sound] * msf.effect_vol * _vol_factor_by_zoom[vp->zoom]) >> 15
 			);
 			return;
@@ -219,7 +224,7 @@
 {
 	StartSound(
 		_sound_idx[sound],
-		4,
+		0,
 		(_sound_base_vol[sound] * msf.effect_vol) >> 7
 	);
 }
