Index: train_cmd.c
===================================================================
--- train_cmd.c	(revision 2945)
+++ train_cmd.c	(working copy)
@@ -113,7 +113,7 @@
 		}
 
 		// max speed is the minimum of the speed limits of all vehicles in the consist
-		if (rvi_u->max_speed != 0)
+		if (rvi_u->max_speed != 0 && !UsesWagonOverride(u))
 			max_speed = min(rvi_u->max_speed, max_speed);
 
 		// check the vehicle length (callback)
